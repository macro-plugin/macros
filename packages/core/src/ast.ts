import type { ArrayExpression, ArrayPattern, ArrowFunctionExpression, AssignmentExpression, AssignmentPattern, AssignmentPatternProperty, AssignmentProperty, AwaitExpression, BigIntLiteral, BinaryExpression, BlockStatement, BooleanLiteral, BreakStatement, CallExpression, CatchClause, ClassDeclaration, ClassExpression, ClassMethod, ClassProperty, ComputedPropName, ConditionalExpression, Constructor, ContinueStatement, DebuggerStatement, Decorator, DoWhileStatement, EmptyStatement, ExportAllDeclaration, ExportDeclaration, ExportDefaultDeclaration, ExportDefaultExpression, ExportDefaultSpecifier, ExportNamedDeclaration, ExportNamespaceSpecifier, ExpressionStatement, ForInStatement, ForOfStatement, ForStatement, FunctionDeclaration, FunctionExpression, GetterProperty, Identifier, IfStatement, Import, ImportDeclaration, ImportDefaultSpecifier, ImportNamespaceSpecifier, Invalid, JSXAttribute, JSXClosingElement, JSXClosingFragment, JSXElement, JSXEmptyExpression, JSXExpressionContainer, JSXFragment, JSXMemberExpression, JSXNamespacedName, JSXOpeningElement, JSXOpeningFragment, JSXSpreadChild, JSXText, KeyValuePatternProperty, KeyValueProperty, LabeledStatement, MemberExpression, MetaProperty, MethodProperty, Module, NamedExportSpecifier, NamedImportSpecifier, NewExpression, NullLiteral, NumericLiteral, ObjectExpression, ObjectPattern, OptionalChainingCall, OptionalChainingExpression, Param, ParenthesisExpression, PrivateMethod, PrivateName, PrivateProperty, RegExpLiteral, RestElement, ReturnStatement, Script, SequenceExpression, SetterProperty, SpreadElement, StaticBlock, StringLiteral, Super, SuperPropExpression, SwitchCase, SwitchStatement, TaggedTemplateExpression, TemplateElement, TemplateLiteral, ThisExpression, ThrowStatement, TryStatement, TsArrayType, TsAsExpression, TsCallSignatureDeclaration, TsConditionalType, TsConstAssertion, TsConstructSignatureDeclaration, TsConstructorType, TsEnumDeclaration, TsEnumMember, TsExportAssignment, TsExpressionWithTypeArguments, TsExternalModuleReference, TsFunctionType, TsGetterSignature, TsImportEqualsDeclaration, TsImportType, TsIndexSignature, TsIndexedAccessType, TsInferType, TsInstantiation, TsInterfaceBody, TsInterfaceDeclaration, TsIntersectionType, TsKeywordType, TsLiteralType, TsMappedType, TsMethodSignature, TsModuleBlock, TsModuleDeclaration, TsNamespaceDeclaration, TsNamespaceExportDeclaration, TsNonNullExpression, TsOptionalType, TsParameterProperty, TsParenthesizedType, TsPropertySignature, TsQualifiedName, TsRestType, TsSetterSignature, TsTemplateLiteralType, TsThisType, TsTupleElement, TsTupleType, TsTypeAliasDeclaration, TsTypeAnnotation, TsTypeAssertion, TsTypeLiteral, TsTypeOperator, TsTypeParameter, TsTypeParameterDeclaration, TsTypeParameterInstantiation, TsTypePredicate, TsTypeQuery, TsTypeReference, TsUnionType, UnaryExpression, UpdateExpression, VariableDeclaration, VariableDeclarator, WhileStatement, WithStatement, YieldExpression } from "@swc/core"

export interface AST {
 "Identifier": Identifier
 "StringLiteral": StringLiteral
 "NumericLiteral": NumericLiteral
 "BigIntLiteral": BigIntLiteral
 "BooleanLiteral": BooleanLiteral
 "NullLiteral": NullLiteral
 "RegExpLiteral": RegExpLiteral
//  "Argument": Argument
 "CallExpression": CallExpression
 "ClassProperty": ClassProperty
 "PrivateProperty": PrivateProperty
 "Param": Param
 "Constructor": Constructor
 "ClassMethod": ClassMethod
 "PrivateMethod": PrivateMethod
 "StaticBlock": StaticBlock
 "Decorator": Decorator
 "FunctionDeclaration": FunctionDeclaration
 "ClassDeclaration": ClassDeclaration
 "VariableDeclaration": VariableDeclaration
 "VariableDeclarator": VariableDeclarator
 "OptionalChainingExpression": OptionalChainingExpression
 "OptionalChainingCall": OptionalChainingCall
 "ThisExpression": ThisExpression
 "ArrayExpression": ArrayExpression
//  "ExprOrSpread": ExprOrSpread
 "ObjectExpression": ObjectExpression
 "SpreadElement": SpreadElement
 "UnaryExpression": UnaryExpression
 "UpdateExpression": UpdateExpression
 "BinaryExpression": BinaryExpression
 "FunctionExpression": FunctionExpression
 "ClassExpression": ClassExpression
 "AssignmentExpression": AssignmentExpression
 "MemberExpression": MemberExpression
 "SuperPropExpression": SuperPropExpression
 "ConditionalExpression": ConditionalExpression
 "Super": Super
 "Import": Import
 "NewExpression": NewExpression
 "SequenceExpression": SequenceExpression
 "ArrowFunctionExpression": ArrowFunctionExpression
 "YieldExpression": YieldExpression
 "MetaProperty": MetaProperty
 "AwaitExpression": AwaitExpression
 "TemplateLiteral": TemplateLiteral
 "TaggedTemplateExpression": TaggedTemplateExpression
 "TemplateElement": TemplateElement
 "ParenthesisExpression": ParenthesisExpression
 "PrivateName": PrivateName
 "JSXMemberExpression": JSXMemberExpression
 "JSXNamespacedName": JSXNamespacedName
 "JSXEmptyExpression": JSXEmptyExpression
 "JSXExpressionContainer": JSXExpressionContainer
 "JSXSpreadChild": JSXSpreadChild
 "JSXOpeningElement": JSXOpeningElement
 "JSXClosingElement": JSXClosingElement
 "JSXAttribute": JSXAttribute
 "JSXText": JSXText
 "JSXElement": JSXElement
 "JSXFragment": JSXFragment
 "JSXOpeningFragment": JSXOpeningFragment
 "JSXClosingFragment": JSXClosingFragment
 "ExportDefaultExpression": ExportDefaultExpression
 "ExportDeclaration": ExportDeclaration
 "ImportDeclaration": ImportDeclaration
 "ExportAllDeclaration": ExportAllDeclaration
 "ExportNamedDeclaration": ExportNamedDeclaration
 "ExportDefaultDeclaration": ExportDefaultDeclaration
 "ImportDefaultSpecifier": ImportDefaultSpecifier
 "ImportNamespaceSpecifier": ImportNamespaceSpecifier
 "NamedImportSpecifier": NamedImportSpecifier
 "ExportNamespaceSpecifier": ExportNamespaceSpecifier
 "ExportDefaultSpecifier": ExportDefaultSpecifier
 "NamedExportSpecifier": NamedExportSpecifier
 "Module": Module
 "Script": Script
 "ArrayPattern": ArrayPattern
 "ObjectPattern": ObjectPattern
 "AssignmentPattern": AssignmentPattern
 "RestElement": RestElement
 "KeyValuePatternProperty": KeyValuePatternProperty
 "AssignmentPatternProperty": AssignmentPatternProperty
 "KeyValueProperty": KeyValueProperty
 "AssignmentProperty": AssignmentProperty
 "GetterProperty": GetterProperty
 "SetterProperty": SetterProperty
 "MethodProperty": MethodProperty
 "ComputedPropName": ComputedPropName
 "BlockStatement": BlockStatement
 "ExpressionStatement": ExpressionStatement
 "EmptyStatement": EmptyStatement
 "DebuggerStatement": DebuggerStatement
 "WithStatement": WithStatement
 "ReturnStatement": ReturnStatement
 "LabeledStatement": LabeledStatement
 "BreakStatement": BreakStatement
 "ContinueStatement": ContinueStatement
 "IfStatement": IfStatement
 "SwitchStatement": SwitchStatement
 "ThrowStatement": ThrowStatement
 "TryStatement": TryStatement
 "WhileStatement": WhileStatement
 "DoWhileStatement": DoWhileStatement
 "ForStatement": ForStatement
 "ForInStatement": ForInStatement
 "ForOfStatement": ForOfStatement
 "SwitchCase": SwitchCase
 "CatchClause": CatchClause
 "TsTypeAnnotation": TsTypeAnnotation
 "TsTypeParameterDeclaration": TsTypeParameterDeclaration
 "TsTypeParameter": TsTypeParameter
 "TsTypeParameterInstantiation": TsTypeParameterInstantiation
 "TsParameterProperty": TsParameterProperty
 "TsQualifiedName": TsQualifiedName
 "TsCallSignatureDeclaration": TsCallSignatureDeclaration
 "TsConstructSignatureDeclaration": TsConstructSignatureDeclaration
 "TsPropertySignature": TsPropertySignature
 "TsGetterSignature": TsGetterSignature
 "TsSetterSignature": TsSetterSignature
 "TsMethodSignature": TsMethodSignature
 "TsIndexSignature": TsIndexSignature
 "TsKeywordType": TsKeywordType
 "TsThisType": TsThisType
 "TsFunctionType": TsFunctionType
 "TsConstructorType": TsConstructorType
 "TsTypeReference": TsTypeReference
 "TsTypePredicate": TsTypePredicate
 "TsImportType": TsImportType
 "TsTypeQuery": TsTypeQuery
 "TsTypeLiteral": TsTypeLiteral
 "TsArrayType": TsArrayType
 "TsTupleType": TsTupleType
 "TsTupleElement": TsTupleElement
 "TsOptionalType": TsOptionalType
 "TsRestType": TsRestType
 "TsUnionType": TsUnionType
 "TsIntersectionType": TsIntersectionType
 "TsConditionalType": TsConditionalType
 "TsInferType": TsInferType
 "TsParenthesizedType": TsParenthesizedType
 "TsTypeOperator": TsTypeOperator
 "TsIndexedAccessType": TsIndexedAccessType
 "TsMappedType": TsMappedType
 "TsLiteralType": TsLiteralType
 "TsTemplateLiteralType": TsTemplateLiteralType
 "TsInterfaceDeclaration": TsInterfaceDeclaration
 "TsInterfaceBody": TsInterfaceBody
 "TsExpressionWithTypeArguments": TsExpressionWithTypeArguments
 "TsTypeAliasDeclaration": TsTypeAliasDeclaration
 "TsEnumDeclaration": TsEnumDeclaration
 "TsEnumMember": TsEnumMember
 "TsModuleDeclaration": TsModuleDeclaration
 "TsModuleBlock": TsModuleBlock
 "TsNamespaceDeclaration": TsNamespaceDeclaration
 "TsImportEqualsDeclaration": TsImportEqualsDeclaration
 "TsExternalModuleReference": TsExternalModuleReference
 "TsExportAssignment": TsExportAssignment
 "TsNamespaceExportDeclaration": TsNamespaceExportDeclaration
 "TsAsExpression": TsAsExpression
//  "TsSatisfiesExpression": TsSatisfiesExpression
 "TsInstantiation": TsInstantiation
 "TsTypeAssertion": TsTypeAssertion
 "TsConstAssertion": TsConstAssertion
 "TsNonNullExpression": TsNonNullExpression
 "Invalid": Invalid
}
