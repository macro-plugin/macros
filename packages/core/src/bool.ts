import { Handler, LabeledMacro, Statement } from "./types";

/**
 * Create a macro plugin that converts `labeled: {}` or `labeled: (...args) => {}` to `$specifier((...args) => {})`,
 * and also `import { $specifier } from $source`
 * @param specifier - the function name
 * @param source - the module path
 * @param allowParams - allow convert the input array to params, default is false.
 * @returns - A labeled macro plugin
 */
export const createLabeledBool: ((id: string) => LabeledMacro) = (id) => {
  return (ast: Statement, code: string, handler: Handler) => {
    // @ts-ignore
    if (ast.type === 'ExpressionStatement' && ast.expression.type == 'BooleanLiteral') {
      // @ts-ignore
      handler.set(id, ast.expression.value);
      return { type: "EmptyStatement" }
    } else {
      throw new Error('this macro only accept a Boolean Literal')
    }
  }
}
