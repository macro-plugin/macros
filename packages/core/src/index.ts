export { createLabeledBlock } from "./block"
export { createLabeledExpr } from "./expr"
export { createSwcPlugin, transform, transformAst, transformAsync } from "./transform"
export { print, printAsync, printExpr, printExprAsync } from "./print"
export { parse, parseAsync, parseExpr, parseType } from "./parse"
export { walk, Walker } from "./walk"
export { macro, $Macro, $LitMacro, $ExprMacro, $TmplMacro, $TypeMacro, $LabeledMacro } from "./macro"
export * from "./api"
export * from "./lib"
export * from "./types"
export * from "./utils"
