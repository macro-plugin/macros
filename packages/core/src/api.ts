import { ArrowFunctionExpression, Expression, FunctionDeclaration, FunctionExpression, Invalid, Param, TsType } from "@swc/core"
import type { BaseNode, ExprMacro, LabeledMacro, MacroPlugin, WalkContext } from "./types"

import { isRegExp } from "./utils"
import { parseExpr } from "./parse"
import { walk } from "./walk"

export function createMacro (macro: MacroPlugin) {
  return macro
}

function createLit (this: WalkContext, value: unknown): BaseNode {
  if (value === undefined) {
    return {
      type: "Identifier",
      span: {
        start: 0,
        end: 0,
        ctxt: 2
      },
      value: "undefined",
      optional: false
    }
  }
  if (value && typeof value === "object") {
    if (isRegExp(value)) return this.parseExpr(value.toString())
    if ("span" in value && "type" in value) return value as BaseNode
  }
  if (typeof value === "function") return this.parseExpr(value.toString())
  return this.parseExpr(JSON.stringify(value))
}

export function createLitMacro(map: Record<string, unknown>): MacroPlugin;
export function createLitMacro<T>(key: string, value: T): MacroPlugin;
export function createLitMacro (arg: string | Record<string, unknown>, value?: unknown): MacroPlugin {
  return createMacro(typeof arg === "string"
    ? {
      Identifier (ast) {
        if (ast.value === arg && !this.track(arg)) return createLit.apply(this, [value])
      }
    }
    : {
      Identifier (ast) {
        if (ast.value in arg && !this.track(ast.value)) return createLit.apply(this, [arg[ast.value]])
      }
    })
}

function flatExpr (f: Function, args: Expression[], typeParams?: TsType[], optional = false): Expression {
  const ast = parseExpr(f.toString()) as FunctionDeclaration | FunctionExpression | ArrowFunctionExpression
  const params: Record<string, { value?: Expression }> = {}

  ast.params.forEach((p, i) => {
    const pat = ast.type === "ArrowFunctionExpression" ? p : ((p as Param).pat)
    if (pat.type === "Identifier") {
      params[pat.value] = args[i] ? { value: args[i] } : {}
    } else if (pat.type === "AssignmentPattern") {
      if (pat.left.type === "Identifier") {
        params[pat.left.value] = { value: args[i] || pat.right }
      }
    }
  })

  let output: Invalid | Expression = {
    type: "Invalid",
    span: {
      start: 0,
      end: 0,
      ctxt: 0
    }
  }

  if (ast.body) {
    walk(ast.body, {
      // @ts-ignore
      enter (ast: BaseNode) {
        if (ast.type === "Identifier" && ast.value in params) {
          const v = params[ast.value].value
          if (v) this.replace(v)
        }
      },
      // @ts-ignore
      leave (ast: BaseNode) {
        if (ast.type === "ReturnStatement" && ast.argument) {
          output = ast.argument
        }
      }
    })

    if (ast.body.type !== "BlockStatement") output = ast.body
  }

  return output
}

export function createExprMacro (name: string, f: Function | ExprMacro | { enter?: ExprMacro, leave?: ExprMacro }): MacroPlugin {
  if (typeof f === "object") {
    return createMacro({
      CallExpression: {
        enter (ast) {
          if (f.enter && ast.callee.type === "Identifier" && ast.callee.value === name && !this.track(ast.callee.value)) {
            return f.enter.apply(this, [ast.arguments.map(i => i.expression), ast.typeArguments?.params, ast.callee.optional])
          }
        },
        leave (ast) {
          if (f.leave && ast.callee.type === "Identifier" && ast.callee.value === name && !this.track(ast.callee.value)) {
            return f.leave.apply(this, [ast.arguments.map(i => i.expression), ast.typeArguments?.params, ast.callee.optional])
          }
        }
      }
    })
  }
  return createMacro({
    CallExpression (ast) {
      if (ast.callee.type === "Identifier" && ast.callee.value === name && !this.track(ast.callee.value)) {
        const args = ast.arguments.map(i => i.expression)
        const tys = ast.typeArguments?.params
        return f.toString().startsWith("function") ? (f as ExprMacro).apply(this, [args, tys, ast.callee.optional]) : flatExpr(f, args, tys, ast.callee.optional)
      }
    }
  })
}

export function createTypeMacro () {

}

export function createTmplMacro (tag: string,) {

}

export function createLabeledMacro (label: string, f: LabeledMacro | {
  enter?: LabeledMacro,
  leave?: LabeledMacro,
}) {
  return createMacro({
    LabeledStatement: {
      enter (ast, parent, prop, index) {
        if (ast.label.value === label) return (typeof f === "object" ? f.enter : f)?.apply(this, [ast.body, parent, prop, index])
      },
      leave (ast, parent, prop, index) {
        if (ast.label.value === label && typeof f === "object") return f.leave?.apply(this, [ast.body, parent, prop, index])
      }
    }
  })
}
