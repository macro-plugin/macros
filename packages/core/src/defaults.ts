import { TsFunctionType, TsType } from "@swc/core"

export const defaultGlobalExpr = {
  type: "TsFunctionType",
  span: {
    start: 199,
    end: 221,
    ctxt: 0
  },
  params: [
    {
      type: "RestElement",
      span: {
        start: 203,
        end: 215,
        ctxt: 0
      },
      rest: {
        start: 5691,
        end: 5694,
        ctxt: 0
      },
      argument: {
        type: "Identifier",
        span: {
          start: 206,
          end: 210,
          ctxt: 4
        },
        value: "args",
        optional: false,
      },
      typeAnnotation: {
        type: "TsTypeAnnotation",
        span: {
          start: 210,
          end: 215,
          ctxt: 0
        },
        typeAnnotation: {
          type: "TsArrayType",
          span: {
            start: 212,
            end: 215,
            ctxt: 0
          },
          elemType: {
            type: "TsTypeReference",
            span: {
              start: 212,
              end: 213,
              ctxt: 0
            },
            typeName: {
              type: "Identifier",
              span: {
                start: 212,
                end: 213,
                ctxt: 4
              },
              value: "T",
              optional: false
            },
          }
        }
      }
    }
  ],
  typeParams: {
    type: "TsTypeParameterDeclaration",
    span: {
      start: 199,
      end: 202,
      ctxt: 0
    },
    parameters: [
      {
        type: "TsTypeParameter",
        span: {
          start: 200,
          end: 201,
          ctxt: 0
        },
        name: {
          type: "Identifier",
          span: {
            start: 200,
            end: 201,
            ctxt: 4
          },
          value: "T",
          optional: false
        },
        in: false,
        out: false,
      }
    ]
  },
  typeAnnotation: {
    type: "TsTypeAnnotation",
    span: {
      start: 217,
      end: 221,
      ctxt: 0
    },
    typeAnnotation: {
      type: "TsTypeReference",
      span: {
        start: 220,
        end: 221,
        ctxt: 0
      },
      typeName: {
        type: "Identifier",
        span: {
          start: 220,
          end: 221,
          ctxt: 4
        },
        value: "T",
        optional: false
      },
    }
  }
} as TsFunctionType

export const defaultGlobalType = {
  type: "TsFunctionType",
  span: {
    start: 182,
    end: 192,
    ctxt: 0
  },
  params: [],
  typeParams: {
    type: "TsTypeParameterDeclaration",
    span: {
      start: 182,
      end: 185,
      ctxt: 0
    },
    parameters: [
      {
        type: "TsTypeParameter",
        span: {
          start: 183,
          end: 184,
          ctxt: 0
        },
        name: {
          type: "Identifier",
          span: {
            start: 183,
            end: 184,
            ctxt: 4
          },
          value: "T",
          optional: false
        },
        in: false,
        out: false,
      }
    ]
  },
  typeAnnotation: {
    type: "TsTypeAnnotation",
    span: {
      start: 188,
      end: 192,
      ctxt: 0
    },
    typeAnnotation: {
      type: "TsTypeReference",
      span: {
        start: 191,
        end: 192,
        ctxt: 0
      },
      typeName: {
        type: "Identifier",
        span: {
          start: 191,
          end: 192,
          ctxt: 4
        },
        value: "T",
        optional: false
      },
    }
  }
} as TsFunctionType

export const defaultGlobalTmpl = {
  type: "TsKeywordType",
  span: {
    start: 236,
    end: 242,
    ctxt: 0
  },
  kind: "string"
} as TsType
