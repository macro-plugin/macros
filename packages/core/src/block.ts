import { ArrowFunctionExpression, ExpressionStatement } from "@swc/core";

import { MacroPlugin } from "./types";
import { createMacro } from "./api";

/**
 * Create a macro plugin that converts `labeled: {}` or `labeled: (...args) => {}` to `$specifier((...args) => {})`,
 * and also `import { $specifier } from $source`
 * @param specifier - the function name
 * @param source - the module path
 * @param allowParams - allow convert the input array to params, default is false.
 * @returns - A labeled macro plugin
 */
export const createLabeledBlock: ((label: string, specifier: string, source: string, allowParams?: boolean) => MacroPlugin) = (label, specifier, source, allowParams = false) => {
  return createMacro({
    LabeledStatement(ast) {
      if (ast.label.value != label) return;
      this.import([{ name: specifier }], source);
      if (ast.body.type == "BlockStatement") {
        return ({
          type: "ExpressionStatement",
          expression: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: specifier,
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            arguments: [
              {
                "expression": {
                  type: 'ArrowFunctionExpression',
                  generator: false,
                  async: false,
                  params: [],
                  body: ast.body,
                  span: {
                    start: 0,
                    end: 0,
                    ctxt: 0
                  }
                } as ArrowFunctionExpression
              }
            ],
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          },
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        }) as ExpressionStatement
      } else if (ast.body.type == "ExpressionStatement" && ast.body.expression.type == 'ArrowFunctionExpression') {
        return ({
          type: "ExpressionStatement",
          expression: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: specifier,
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            arguments: [
              {
                expression: ast.body.expression
              }
            ],
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          },
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        }) as ExpressionStatement
      } else if (allowParams && ast.body.type == "ExpressionStatement" && ast.body.expression.type == 'ArrayExpression') {
        return ({
          type: "ExpressionStatement",
          expression: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: specifier,
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            arguments: ast.body.expression.elements,
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          },
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        }) as ExpressionStatement
      } else {
        throw new Error('this macro only accept a ArrowFunction or BlockStatement' + (allowParams ? ' or an ArrayExpression' : ''))
      }
    }
  })
}
