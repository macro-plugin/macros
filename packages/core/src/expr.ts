import { CallExpression } from "@swc/core"
import { MacroPlugin } from "./types"
import { createMacro } from "./api"

export const createLabeledExpr: ((label: string, specifier: string, source: string) => MacroPlugin) = (label, specifier, source) => {
  return createMacro({
    LabeledStatement (ast) {
      if (ast.label.value != label) return
      if (ast.body.type == "ExpressionStatement") {
        this.import([{ name: specifier }], source)

        ast.body.expression = {
          type: "CallExpression",
          span: {
            start: 109,
            end: 119,
            ctxt: 0
          },
          callee: {
            type: "Identifier",
            span: {
              start: 109,
              end: 112,
              ctxt: 1
            },
            value: specifier,
            optional: false
          },
          arguments: [
            {
              expression: ast.body.expression
            }
          ],
        } as CallExpression

        return ast.body
      } else {
        throw new Error("this macro only accept an Expression")
      }
    }
  })
}
