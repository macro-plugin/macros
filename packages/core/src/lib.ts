import { Expression, Identifier } from "@swc/core"
import { createExprMacro, createTmplMacro } from "./api"
import { evalAst, evalExpr, span } from "./utils"

import { printExpr } from "./print"
import { walk } from "./walk"

export var $Eval = createExprMacro("$Eval", function (args) {
  const t = args[0].type

  if (t === "StringLiteral") return this.parseExpr(evalExpr(args[0].value))
  if (t === "TemplateLiteral") {
    const track = (name: string) => this.track(name)
    walk(args[0], {
      enter (node) {
        if (node.type === "Identifier") {
          const v = track((node as Identifier).value)
          if (v && v.value) this.replace(v.value)
        }
      }
    })

    return this.parseExpr(evalExpr(evalExpr(this.printExpr(args[0]))))
  }

  if (t === "FunctionExpression" || t === "ArrowFunctionExpression") return this.parseExpr(evalExpr(`(${this.printExpr(args[0])})(${args.slice(1).map(i => this.printExpr(i)).join(",")})`))
  return this.parseExpr(evalAst(args[0]).toString())
}, "(<T>(expr: string) => T) & (<T>(expr: T) => T) & (<F extends (...args: any) => any>(expr: F, ...args: Parameters<F>) => ReturnType<F>)")

export const printAst = (ast: object) => JSON.stringify(ast).replace(/("(start|end|ctxt)":\s*)(\d+)/g, "$10")

export var $Ast = createExprMacro("$Ast", function (args) {
  return this.parseExpr(printAst(args[0].type === "StringLiteral" ? this.parseExpr(args[0].value) : args[0]))
}, '<T>(expr: T) => import("@swc/core").Expression')

export var $Env = createExprMacro("$Env", function (args, typeArgs) {
  if (args[0].type !== "StringLiteral") throw new Error("$Env macro only support StringLiteral as input.")
  const value = process.env[args[0].value] ?? ""

  if (typeArgs && typeArgs?.[0].type === "TsKeywordType") {
    switch (typeArgs[0].kind) {
      case "boolean":
        return {
          type: "BooleanLiteral",
          value: !!value,
          span
        }
      case "number":
        return {
          type: "NumericLiteral",
          value: +value,
          span
        }
    }
  }

  return {
    type: "StringLiteral",
    span,
    value
  }
}, "<R>(key: string) => R")

export var $Stringify = createExprMacro("$Stringify", function (args, typeArgs) {
  return {
    type: "StringLiteral",
    span,
    value: typeArgs && typeArgs.length > 0 ? this.printType(typeArgs[0]) : this.printExpr(args[0])
  }
}, "((expr: any) => string) & (<T>() => string)")

export const printTmpl = (strings: string[], exprs: Expression[]) => strings.reduce((query, queryPart, i) => {
  const valueExists = i < exprs.length
  const text = query + queryPart

  return valueExists ? text + printExpr(exprs[i]).code : text
}, "")

export const printRawTmpl = (strings: string[], exprs: unknown[]) => strings.reduce((query, queryPart, i) => {
  const valueExists = i < exprs.length
  const text = query + queryPart

  return valueExists ? text + exprs[i] : text
}, "")

function walkObject (obj: object | null, onEnter: (k: string, v: unknown, parent: Record<string, any>) => unknown, grand?: Record<string, any>, grandKey?: string) {
  if (obj && typeof obj === "object") {
    const allKeys = Object.keys(obj)
    for (let i = 0; i < allKeys.length; i++) {
      const k = allKeys[i]
      const v = obj[k as keyof typeof obj]

      const r = onEnter(k, v, obj)
      if (r != null && grand) {
        grand[grandKey!] = r
        break
      }

      if (typeof v === "object") {
        walkObject(v, onEnter, obj, k)
      }
    }
  }
  return obj
}

export var $Expr = createTmplMacro("$Expr", function (strings, ...exprs) {
  const exprMarkers: string[] = exprs.map((_, i) => "_macro_marker_" + i + "_")
  const options = { syntax: "typescript", tsx: true } as const
  const ast = this.parseExpr(printRawTmpl(strings, exprMarkers), options)

  walkObject(ast, (k, v, parent) => {
    if (k === "span") {
      parent[k] = span
      return
    }

    if (typeof v === "string") {
      const i = exprMarkers.findIndex((m) => m === v)
      if (i !== -1) return "__macro$$Start__" + this.printExpr(exprs[i]) + "__macro$$End__"
    }
  })

  const expr = JSON.stringify(ast, undefined, 2).replace(/("__macro\$\$Start__)|(__macro\$\$End__")/g, "").replace(/\\"/g, '"')

  return this.parseExpr(expr)
}, "import(\"@swc/core\").Expression")

export var $Quote = createTmplMacro("$Quote", function (strings, ...exprs) {
  const exprMarkers: string[] = exprs.map((_, i) => "_macro_marker_" + i + "_")
  const options = { syntax: "typescript", tsx: true } as const
  const ast = this.parse(printRawTmpl(strings, exprMarkers), options)

  walkObject(ast, (k, v, parent) => {
    if (k === "span") {
      parent[k] = span
      return
    }

    if (typeof v === "string") {
      const i = exprMarkers.findIndex((m) => m === v)
      if (i !== -1) return "__macro$$Start__" + this.printExpr(exprs[i]) + "__macro$$End__"
    }
  })

  const expr = JSON.stringify(ast, undefined, 2).replace(/("__macro\$\$Start__)|(__macro\$\$End__")/g, "").replace(/\\"/g, '"')

  return this.parseExpr(expr)
}, "(import(\"@swc/core\").Expression)[]")
