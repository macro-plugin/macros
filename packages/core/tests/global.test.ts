import { createMacro, transform } from "../src"

import type { VariableDeclaration } from "@swc/core"

const arrow = createMacro({
  FunctionDeclaration (ast) {
    const children = []
    let isArrow = false

    if (!ast.body) return

    for (const s of ast.body.stmts || []) {
      if (s.type === "LabeledStatement" && s.body.type === "ExpressionStatement") {
        const expr = s.body.expression
        if (s.label.value === "arrow") {
          if (expr.type === "BooleanLiteral" && expr.value) {
            isArrow = true
          }
          continue
        }
      }
      children.push(s)
    }

    ast.body.stmts = children

    if (isArrow) {
      return {
        type: "VariableDeclaration",
        kind: "const",
        declare: false,
        span: {
          start: 0,
          end: 0,
          ctxt: 0
        },
        declarations: [
          {
            type: "VariableDeclarator",
            id: ast.identifier,
            init: {
              type: "ArrowFunctionExpression",
              generator: false,
              async: false,
              params: [],
              body: ast.body,
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            definite: false,
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          }
        ]
      } as VariableDeclaration
    }
  }
})

const inject = createMacro(function () {
  this.import([{ name: "test", kind: "default" }], "test")
  this.import([{ name: "ref" }], "vue")
})

test("transform function block", () => {
  const code = `
    function add(a, b) {
      arrow: true
      return a + b
    }
  `

  const code2 = `
    function add(a, b) {
      arrow: false
      return a + b
    }
  `
  expect(transform(code, { plugins: [arrow] }).code).toMatchSnapshot()
  expect(transform(code2, { plugins: [arrow] }).code).toMatchSnapshot()
})

test("inject imports", () => {
  const code = `
  function myComponent() {
    const count = ref(0);
  }
  `

  expect(transform(code, {
    plugins: [inject],
    jsc: {
      parser: {
        syntax: "typescript",
      },
    },
  }).code).toMatchSnapshot()
})
