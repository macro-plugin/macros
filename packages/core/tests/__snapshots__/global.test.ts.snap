// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inject imports 1`] = `
"import test from "test";
import { ref } from "vue";
function myComponent() {
    const count = ref(0);
}
"
`;

exports[`transform function block 1`] = `
"const add = ()=>{
    return a + b;
};
"
`;

exports[`transform function block 2`] = `
"function add(a, b) {
    return a + b;
}
"
`;
